src/core/qgsvectorlayerutils.cpp

```diff
@@ -522,6 +523,135 @@ std::unique_ptr<QgsVectorLayerFeatureSource> QgsVectorLayerUtils::getFeatureSour
+{
+  QgsWkbTypes::Type inputWkbType( layer.wkbType( ) );
+  QgsFeatureList resultFeatures;
+  for ( const QgsFeature &f : features )
+  {
+    QgsFeature newF( f );
+    // Fix attributes
+    QgsVectorLayerUtils::matchAttributesToFields( newF, layer.fields( ) );
+    // Does geometry need tranformations?
+    QgsWkbTypes::GeometryType newFGeomType( QgsWkbTypes::geometryType( newF.geometry().wkbType() ) );
+    bool newFHasGeom = newFGeomType !=
+                       QgsWkbTypes::GeometryType::UnknownGeometry &&
+                       newFGeomType != QgsWkbTypes::GeometryType::NullGeometry;
+    bool layerHasGeom = inputWkbType !=
+                        QgsWkbTypes::Type::NoGeometry &&
+                        inputWkbType != QgsWkbTypes::Type::Unknown;
+    // Drop geometry if layer is geometry-less
+    if ( newFHasGeom && ! layerHasGeom )
+    {
+      QgsFeature _f = QgsFeature( layer.fields() );
+      _f.setAttributes( newF.attributes() );
+      resultFeatures.append( _f );
+      continue;  // Skip the rest
+    }
+    // Geometry need fixing
+    if ( newFHasGeom && layerHasGeom && newF.geometry().wkbType() != inputWkbType )
+    {
+      // Single -> multi
+      if ( QgsWkbTypes::isMultiType( inputWkbType ) && ! newF.geometry().isMultipart( ) )
+      {
+        QgsGeometry newGeom( newF.geometry( ) );
+        newGeom.convertToMultiType();
+        newF.setGeometry( newGeom );
+      }
+      // Drop Z/M
+      if ( newF.geometry().constGet()->is3D() && ! QgsWkbTypes::hasZ( inputWkbType ) )
+      {
+        QgsGeometry newGeom( newF.geometry( ) );
+        newGeom.get()->dropZValue();
+        newF.setGeometry( newGeom );
+      }
+      if ( newF.geometry().constGet()->isMeasure() && ! QgsWkbTypes::hasM( inputWkbType ) )
+      {
+        QgsGeometry newGeom( newF.geometry( ) );
+        newGeom.get()->dropMValue();
+        newF.setGeometry( newGeom );
+      }
+      // Add Z/M back, set to 0
+      if ( ! newF.geometry().constGet()->is3D() && QgsWkbTypes::hasZ( inputWkbType ) )
+      {
+        QgsGeometry newGeom( newF.geometry( ) );
+        newGeom.get()->addZValue( 0.0 );
+        newF.setGeometry( newGeom );
+      }
+      if ( ! newF.geometry().constGet()->isMeasure() && QgsWkbTypes::hasM( inputWkbType ) )
+      {
+        QgsGeometry newGeom( newF.geometry( ) );
+        newGeom.get()->addMValue( 0.0 );
+        newF.setGeometry( newGeom );
+      }
+      // Multi -> single
+      if ( ! QgsWkbTypes::isMultiType( inputWkbType ) && newF.geometry().isMultipart( ) )
+      {
+        QgsGeometry newGeom( newF.geometry( ) );
+        const QgsGeometryCollection *parts( static_cast< const QgsGeometryCollection * >( newGeom.constGet() ) );
+        for ( int i = 0; i < parts->partCount( ); i++ )
+        {
+          QgsGeometry g( parts->geometryN( i )->clone() );
+          QgsAttributeMap attrMap;
```

Comment:
This should be pulled out of the part loop, `attrMap` can be reused for every part.